ircd/channel.c

```diff
@@ -716,12 +716,25 @@ can_join(struct Client *source_p, struct Channel *chptr, const char *key, const
 		goto finish_join_check;
 	}
 
-	if(*chptr->mode.key && (EmptyString(key) || irccmp(chptr->mode.key, key)))
+	matchset_for_client(source_p, &ms);
+
+	RB_DLINK_FOREACH(ptr, chptr->invexlist.head)
+	{
+		invex = ptr->data;
+		if (matches_mask(&ms, invex->banstr) ||
+				match_extban(invex->banstr, source_p, chptr, CHFL_INVEX))
+			break;
+	}
+
+	if(*chptr->mode.key && !invex &&
```

Comment:
`invex` will always be valid here, thus always allowing people in.  You need to `NULL` it at the end of every loop above, so that the break skips that and thus it's only valid if it matched.